src/Workspaces/Remote/Core/SolutionAssetStorage.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Serialization;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// Stores solution snapshots available to remote services.
+    /// </summary>
+    internal partial class SolutionAssetStorage
+    {
+        private static int s_scopeId = 1;
+
+        /// <summary>
+        /// Map from solution checksum scope id to its associated <see cref="SolutionState"/>.
+        /// </summary>
+        private readonly ConcurrentDictionary<int, SolutionState> _solutionStates;
+
+        public SolutionAssetStorage()
+        {
+            _solutionStates = new ConcurrentDictionary<int, SolutionState>(concurrencyLevel: 2, capacity: 10);
+        }
+
+        /// <summary>
+        /// Adds given snapshot into the storage. This snapshot will be available within the returned <see cref="Scope"/>.
+        /// </summary>
+        internal async ValueTask<Scope> StoreAssetsAsync(Solution solution, CancellationToken cancellationToken)
+        {
+            var solutionState = solution.State;
+            var solutionChecksum = await solutionState.GetChecksumAsync(cancellationToken).ConfigureAwait(false);
+
+            var id = Interlocked.Increment(ref s_scopeId);
+            var solutionInfo = new PinnedSolutionInfo(
+                id,
+                solutionState.BranchId == solutionState.Workspace.PrimaryBranchId,
+                solutionState.WorkspaceVersion,
+                solutionChecksum);
+
+            Contract.ThrowIfFalse(_solutionStates.TryAdd(id, solutionState));
+
+            return new Scope(this, solutionInfo);
+        }
+
+        /// <summary>
+        /// Retrieve asset of a specified <paramref name="checksum"/> available within <paramref name="scopeId"/> scope from the storage.
+        /// </summary>
+        public async ValueTask<SolutionAsset?> GetAssetAsync(int scopeId, Checksum checksum, CancellationToken cancellationToken)
+        {
+            if (checksum == Checksum.Null)
+            {
+                // check nil case
+                return SolutionAsset.Null;
```

Comment:
ok.  i think i get why you didn't have to write it.  that said, can you doc the writing side.